src/pkg/artifact/dao/dao.go

```diff
@@ -0,0 +1,114 @@
+	// Delete the artifact specified by ID
+	Delete(ctx context.Context, id int64) (err error)
+	// Update updates the artifact. Only the properties specified by "props" will be updated if it is set
+	Update(ctx context.Context, artifact *Artifact, props ...string) (err error)
+	// CreateReference creates the artifact reference
+	CreateReference(ctx context.Context, reference *ArtifactReference) (id int64, err error)
+	// ListReferences lists the artifact references according to the query
+	ListReferences(ctx context.Context, query *q.Query) (references []*ArtifactReference, err error)
+	// DeleteReferences deletes the references referenced by the artifact specified by parent ID
+	DeleteReferences(ctx context.Context, parentID int64) (err error)
+}
+
+// New returns an instance of the default DAO
+func New() DAO {
+	return &dao{}
+}
+
+type dao struct{}
+
+func (d *dao) Count(ctx context.Context, query *q.Query) (int64, error) {
+	return orm.GenerateQuerySetter(ctx, &Artifact{}, query, false).Count()
+}
+func (d *dao) List(ctx context.Context, query *q.Query) ([]*Artifact, error) {
+	artifacts := []*Artifact{}
+	if _, err := orm.GenerateQuerySetter(ctx, &Artifact{}, query, true).All(&artifacts); err != nil {
+		return nil, err
+	}
+	return artifacts, nil
+}
+func (d *dao) Get(ctx context.Context, id int64) (*Artifact, error) {
+	artifact := &Artifact{
+		ID: id,
+	}
+	if err := orm.GetOrmer(ctx).Read(artifact); err != nil {
+		if e, ok := orm.IsNotFoundError(err, "artifact %d not found", id); ok {
+			err = e
+		}
+		return nil, err
+	}
+	return artifact, nil
+}
+func (d *dao) Create(ctx context.Context, artifact *Artifact) (int64, error) {
+	id, err := orm.GetOrmer(ctx).Insert(artifact)
+	if e, ok := orm.IsConflictError(err, "artifact %s already exists under the repository %d",
+		artifact.Digest, artifact.RepositoryID); ok {
+		err = e
+	}
+	return id, err
+}
+func (d *dao) Delete(ctx context.Context, id int64) error {
+	_, err := orm.GetOrmer(ctx).Delete(&Artifact{
+		ID: id,
+	})
+	return err
+}
+func (d *dao) Update(ctx context.Context, artifact *Artifact, props ...string) error {
+	_, err := orm.GetOrmer(ctx).Update(artifact, props...)
+	return err
+}
+func (d *dao) CreateReference(ctx context.Context, reference *ArtifactReference) (int64, error) {
+	id, err := orm.GetOrmer(ctx).Insert(reference)
+	if e, ok := orm.IsConflictError(err, "reference already exists, parent artifact ID: %d, child artifact ID: %d",
+		reference.ParentID, reference.ChildID); ok {
+		err = e
+	}
+	return id, err
+}
+func (d *dao) ListReferences(ctx context.Context, query *q.Query) ([]*ArtifactReference, error) {
+	references := []*ArtifactReference{}
+	if _, err := orm.GenerateQuerySetter(ctx, &ArtifactReference{}, query, true).All(&references); err != nil {
+		return nil, err
+	}
+	return references, nil
+}
+func (d *dao) DeleteReferences(ctx context.Context, parentID int64) error {
+	_, err := orm.GenerateQuerySetter(ctx, &ArtifactReference{}, &q.Query{
```

Comment:
is it appropriate to completely ignore the case when `parentID` does not exist?

Reply:
Checked